lib/utils/remote-build.ts

```diff
@@ -0,0 +1,219 @@
+/*
+Copyright 2016-2017 Resin.io
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import * as JSONStream from 'JSONStream';
+import * as request from 'request';
+import { ResinSDK } from 'resin-sdk';
+import * as Stream from 'stream';
+
+import { tarDirectory } from './compose';
+
+const DEBUG_MODE = !!process.env.DEBUG;
+
+const CURSOR_METADATA_REGEX = /([a-z]+)([0-9]+)?/;
+const TRIM_REGEX = /\n+$/;
+
+export interface RemoteBuild {
+	app: string;
+	owner: string;
+	source: string;
+	auth: string;
+	baseUrl: string;
+
+	sdk: ResinSDK;
+
+	// For internal use
+	releaseId?: number;
+}
+
+interface BuilderMessage {
+	message: string;
+	type?: string;
+	replace?: boolean;
+	// These will be set when the type === 'metadata'
+	resource?: string;
+	value?: string;
+}
+
+async function getBuilderEndpoint(
+	baseUrl: string,
+	owner: string,
+	app: string,
+): Promise<string> {
+	const querystring = await import('querystring');
+	const args = querystring.stringify({ owner, app });
+	return `https://builder.${baseUrl}/v3/build?${args}`;
+}
+
+export async function startRemoteBuild(build: RemoteBuild): Promise<void> {
+	const Bluebird = await import('bluebird');
+
+	return new Bluebird(async (resolve, reject) => {
+		const { stream, uploadSpinner } = await getRequestStream(build);
+		let spinnerStopped = false;
+
+		// Special windows handling (win64 also reports win32)
+		if (process.platform === 'win32') {
+			const readline = (await import('readline')).createInterface({
+				input: process.stdin,
+				output: process.stdout,
+			});
+
+			readline.on('SIGINT', () => process.emit('SIGINT'));
+		}
+
+		// Setup interrupt handlers so we can cancel the build if the user presses
+		// ctrl+c
+
+		// This is necessary because the `exit-hook` module is used by several
+		// dependencies, and will exit without calling the following handler.
+		// Once https://github.com/resin-io/resin-cli/issues/867 has been solved,
+		// we are free to (and definitely should) remove the below line
+		process.removeAllListeners('SIGINT');
+		process.on('SIGINT', () => {
+			process.stderr.write('Received SIGINT, cleaning up. Please wait.\n');
+			cancelBuildIfNecessary(build).then(() => {
+				stream.end();
+				process.exit(130);
+			});
+		});
+
+		const dataHandler = getBuilderMessageHandler(build);
+		stream.on('data', (data: BuilderMessage) => {
+			if (!spinnerStopped) {
+				uploadSpinner.stop();
+				spinnerStopped = true;
```

Comment:
This spinner actually has its own check to make `stop()` safe: https://github.com/resin-io-modules/resin-cli-visuals/blob/edeb543a5c83cb67b684564f7835e8a69dfeaf6a/lib/widgets/spinner.coffee#L78

You could drop `spinnerStopped` entirely, and just call `uploadSpinner.stop()` every time.